rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,261 @@
+`reflect_a` from the previous examples.
+
+
+#### `Dynamic***`
+
+The structures holding generic structural information about some proxied type. 
+* `DynamicMap`
+* `DynamicList`
+* `DynamicTuple`
+* `DynamicStruct`
+* `DynamicTupleStruct`
+
+#### Reflect subtraits
+
+The top level traits in `bevy_reflect`: `Struct`, `TupleStruct`, `Tuple`,
+`List` and `Map`. They are all subtraits of `Reflect`.
+
+## Motivation
+
+The `Reflect` trait is a trap, it doesn't work as expected due to `Dynamic***`
+mixing up with the "regular" types under the `dyn Reflect` trait object.
+
+To solve this, we do not allow more than a single underlying type to "pass for"
+a concrete type in `Reflect`. This implies that the various `Dynamic***` types
+shouldn't implement `Reflect`.
+
+When the underlying type of a `Box<dyn Reflect>` is a `Dynamic***`, a lot of
+the `Reflect` methods become invalid, such as `any`, `reflect_hash`,
+`reflect_partial_eq`, `serializable`, `downcast`, `is` or `take`. However,
+those methods are provided regardless.
+
+### What's under the `dyn Reflect`?
+
+Currently, `Reflect` has very a confusing behavior. Most of the methods on
+`Reflect` are highly dependent on the underlying type, regardless of whether
+they are supposed to stand for the requested type or not.
+
+For example
+
+```rust
+let a = MysteryType::new(); // suppose MysteryType implements Reflect
+let reflect_a: Box<dyn Reflect> = Box::new(a.clone()) as Box<dyn Reflect>;
+let dynamic_a: Box<dyn Reflect> = a.clone_value();
+// This more or less always panic
+assert_eq!((&*reflect_a).type_id(), (&*dynamic_a).type_id())
+```
+
+This is because multiple different things can pretend to be the same thing. In
+itself this shouldn't be an issue, but the current `Reflect` API is leaky, and
+you need to be aware of both this limitation and the underlying type to be able
+to use the `Reflect` API correctly.
+
+
+### `reflect_hash` and co. with `Dynamic***`
+
+The problem is not limited to `type_id`, but also extends to the `reflect_*`
+methods.
+
+```rust
+let a = MysteryType::new(); // suppose MysteryType implements Reflect
+let reflect_a: Box<dyn Reflect> = Box::new(a.clone()) as Box<dyn Reflect>;
+let dynamic_a: Box<dyn Reflect> = a.clone_value();
+let reflect_hash = reflect_a.reflect_hash();
+let dynamic_hash = dynamic_a.reflect_hash();
+// This more or less always panic if `MysteryType` implements `Hash`
+assert_eq!(reflect_hash, dynamic_hash)
+```
+
+
+## Proposition
+
+The problem stems from the duplicity of `dyn Reflect`. We want to rely on the
+underlying methods of the same thing to be unique.
+
+To solve this, we do not allow more than a single underlying type to "pass for"
+a concrete type in `Reflect`. This implies that the various `Dynamic***` types
+shouldn't implement `Reflect`.
+
+## User-facing explanation
+
+* `Reflect` maps 1-to-1 with an underlying type meaning that if a
+  `Box<dyn Reflect>` cannot be downcasted to `T`, you cannot build a `T` from
+  it.
```

Comment:
Won't that prevent `Reflect` from working for non-rust types like in case of scripting?

Reply:
Could you elaborate. How do you use a non-rust type in rust? Wouldn't it need a conversion at one point?

Reply:
The idea would be that you have for example a `PyObject` which is how any python object is internally represented and then it's `Reflect` implementation would present a different type depending on the type the `PyObject` has on the python side. For example for

```python
class Foo(object):
    __slots__ = ('a', 'b')

class Bar(object):
    __slots__ = ('c', 'd')
```

a `PyObject` holding a `Foo` would present the fields `a` and `b`, while a `PyObject` holding a `Bar` would present the fields `c` and `d` and a `PyObject` holding an array would present itself as an array through `Reflect`.

Reply:
But this only exists on the Python side, right? What would this map to in Rust currently if there's no matching struct?

Reply:
It doesn't need to map to a rust type I think. `downcast` could return None for any type you attempt to downcast to. `type_name` would give the python name. It shouldn't map to `DynamicStruct` as that would lose the python type.

Reply:
What's the actual Rust type though? What are you passing around? Could _that_ type be reflected?

Like `PyObject` -To Rust-> `PyObjectContainer` -Reflected-> `dyn Reflect` -Downcast-> `PyObjectContainer` -To Python-> `PyObject`.

Sorry if I’m misunderstanding 😅

Reply:
That would work, but it would have the exact same issue as what the `Dynamic*` types have right now in that the downcasted type doesn't match the semantic type that determines the layout which `.reflect()` gives.

Reply:
Yeah I think I get the issue here. What about the proposition of moving `reflect_ref` and `reflect_mut` to their own trait independent from `Reflect` (call it the `Bavy` trait) and also move all the `Reflect` subtraits to `Bavy`? You could then implement `Bavy` for `PyObject` (or a wrapper type, per orphan rule). Maybe also move the `sidecast` proposition and implement `Bavy` for `ReflectProxy`?

Reply:
I would rather go the opposite way and move all methods except `reflect_ref` and `reflect_mut` into separate traits as not every type may have an implementation for them and `reflect_ref` and `reflect_mut` are the core reflection api.

Reply:
@bjorn3  I think this is a good idea. Maybe I could revise this RFC to have a `UniqueReflect` vs `Reflect`?

Reply:
why would you want to reuse Reflect as a trait for this as opposed to writing specific code for Python? what's the use case? I thought Reflect as a trait is supposed to model how you can represent data in Rust, not in any language ever.